HLTrigger/Egamma/plugins/HLTEgammaCombMassFilter.cc

```diff
@@ -43,12 +45,75 @@ void HLTEgammaCombMassFilter::fillDescriptions(edm::ConfigurationDescriptions& d
+  using namespace trigger;
+  if (saveTags()) {
+    filterproduct.addCollectionTag(l1EGTag_);
+  }
 
-  //trigger::TriggerObjectType firstLegTrigType;
-  std::vector<math::XYZTLorentzVector> firstLegP4s;
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> PrevFilterOutput;
+  iEvent.getByToken(firstLegLastFilterToken_, PrevFilterOutput);
+
+  std::vector<edm::Ref<reco::RecoEcalCandidateCollection> > phoCandsPrev;
+  PrevFilterOutput->getObjects(trigger::TriggerPhoton, phoCandsPrev);
+  std::vector<edm::Ref<reco::RecoEcalCandidateCollection> > clusCandsPrev;
+  PrevFilterOutput->getObjects(trigger::TriggerCluster, clusCandsPrev);
+  std::vector<edm::Ref<reco::ElectronCollection> > eleCandsPrev;
+  PrevFilterOutput->getObjects(trigger::TriggerElectron, eleCandsPrev);
+
+  if (!phoCandsPrev.empty()) {
+    std::vector<bool> save_pho(phoCandsPrev.size(), true);
+    for (unsigned int jj = 0; jj < phoCandsPrev.size(); jj++) {
+      auto const& phoCand1 = phoCandsPrev[jj];
+      for (unsigned int ii = jj + 1; ii < phoCandsPrev.size(); ii++) {
+        auto const& phoCand2 = phoCandsPrev[ii];
+        math::XYZTLorentzVector tempPair = phoCand1->p4() + phoCand2->p4();
+
+        if (tempPair.M() >= minMass_) {
+          for (auto const idx : {jj, ii}) {
+            if (save_pho[idx])
+              filterproduct.addObject(trigger::TriggerPhoton, phoCandsPrev[idx]);
+            save_pho[idx] = false;
+          }
+        }
+      }
+    }
+  } else if (!clusCandsPrev.empty()) {
+    std::vector<bool> save_clus(clusCandsPrev.size(), true);
+    for (unsigned int jj = 0; jj < clusCandsPrev.size(); jj++) {
+      auto const& clusCand1 = clusCandsPrev[jj];
+      for (unsigned int ii = jj + 1; ii < clusCandsPrev.size(); ii++) {
+        auto const& clusCand2 = clusCandsPrev[ii];
+        math::XYZTLorentzVector tempPair = clusCand1->p4() + clusCand2->p4();
+
+        if (tempPair.M() >= minMass_) {
+          for (auto const idx : {jj, ii}) {
+            if (save_clus[idx])
+              filterproduct.addObject(trigger::TriggerCluster, clusCandsPrev[idx]);
+            save_clus[idx] = false;
+          }
+        }
+      }
+    }
+  } else if (!eleCandsPrev.empty()) {
+    std::vector<bool> save_ele(eleCandsPrev.size(), true);
+    for (unsigned int jj = 0; jj < eleCandsPrev.size(); jj++) {
+      auto const& eleCand1 = eleCandsPrev[jj];
+      for (unsigned int ii = jj + 1; ii < eleCandsPrev.size(); ii++) {
+        auto const& eleCand2 = eleCandsPrev[ii];
+        math::XYZTLorentzVector tempPair = eleCand1->p4() + eleCand2->p4();
+
+        if (tempPair.M() >= minMass_) {
+          for (auto const idx : {jj, ii}) {
+            if (save_ele[idx])
+              filterproduct.addObject(trigger::TriggerCluster, eleCandsPrev[idx]);
+            save_ele[idx] = false;
+          }
+        }
+      }
+    }
+  }
 
```

Comment:
This whole added code block seems to depend only on the elements of the first leg (and making pairs from those), not on the elements of the second leg. It is used to fill the filter objects. But below the filter result checks all pairs with one leg from the first leg and one leg from the second leg.

So there is a mismatch of the pairs whose objetcs are recorded vs the pairs whihc are evaluated to find a passing pair.

Reply:
Thanks for pointing our the problem. Although it seems that the current existing filters which have used this  `HLTEgammaCombMassFilter` are all with identical first and second leg, I think it might be better to not enforce this equality so that we can accommodate the possibility of future filters that have different legs. I will try to fix it and ensure their consistency.

Reply:
Fixed in the latest commit